meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+showContract (Both contract1 contract2) =
+  joinWith "" ["Both (",
+               showContract contract1, ") (",
+               showContract contract2, ")"]
+showContract (Choice observation contract1 contract2) =
+  joinWith "" ["Choice (",
+               showObservation observation, ") (",
+               showContract contract1, ") (",
+               showContract contract2, ")"]
+showContract (When observation timeout contract1 contract2) =
+  joinWith "" ["When (",
+               showObservation observation, ") ",
+               toString timeout, " (",
+               showContract contract1, ") (",
+               showContract contract2, ")"]
+showContract (While observation timeout contract1 contract2) =
+  joinWith "" ["While (",
+               showObservation observation, ") ",
+               toString timeout, " (",
+               showContract contract1, ") (",
+               showContract contract2, ")"]
+showContract (Scale value1 value2 value3 contract) =
+  joinWith "" ["Scale (",
+               showValue value1, ") (",
+               showValue value2, ") (",
+               showValue value3, ") (",
+               showContract contract, ")"]
+showContract (Let letLabel contract1 contract2) =
+  joinWith "" ["Let ",
+               toString letLabel, " (",
+               showContract contract1, ") (",
+               showContract contract2, ")"]
+showContract (Use letLabel) =
+  joinWith "" ["Use ",
+               toString letLabel]
+
+-- Data type for Inputs with their information 
+data Input = IChoice IdChoice Choice
+           | IOracle IdOracle BlockNumber BigInt
+derive instance eqInput :: Eq Input  
+derive instance ordInput :: Ord Input  
+
+data AnyInput = Action IdAction
+              | Input Input
+derive instance eqAnyInput :: Eq AnyInput  
+derive instance ordAnyInput :: Ord AnyInput  
+
+data IdInput = IdChoice IdChoice
+             | IdOracle IdOracle
+derive instance eqIdInput :: Eq IdInput  
+derive instance ordIdInput :: Ord IdInput  
+
+
+type TimeoutData = M.Map Timeout (S.Set IdCommit)
+
+type CommitInfoRecord = { person :: Person
+                        , amount :: BigInt
+                        , timeout :: Timeout }
+
+data CommitInfo = CommitInfo { redeemedPerPerson :: M.Map Person BigInt
+                             , currentCommitsById :: M.Map IdCommit CommitInfoRecord 
+                             , expiredCommitIds :: S.Set IdCommit
+                             , timeoutData :: TimeoutData }
+--derive instance eqCommitInfo :: Eq CommitInfo  
+--derive instance ordCommitInfo :: Ord CommitInfo  
+
+type OracleDataPoint = { blockNumber :: BlockNumber
+                       , value :: BigInt }
+
+data State = State { commits :: CommitInfo
+                   , choices :: M.Map WIdChoice Choice
+                   , oracles :: M.Map IdOracle OracleDataPoint
+                   , usedIds :: S.Set IdAction}
+--derive instance eqState :: Eq State  
+--derive instance ordState :: Ord State  
```

Comment:
why comment out these?

Reply:
They didn't work with the version of the Prelude that is in the bower. Some problem deriving instances from records. And I didn't manage to upgrade the Prelude, seems quite tricky to get dependencies right.

Reply:
yeah, at this stage we should avoid changing dependencies. At some point we need to upgrade to PS 0.12 and possibly use spago instead of bower so we don't have this issue in the future however it is a big job as some libraries that we use aren't ready